### PR TITLE
Refactor listview to always focus the row

### PR DESCRIPTION
**Pull request by LFDanLu**
Closes https://github.com/adobe/react-spectrum/issues/2946
Alternate approach from https://github.com/adobe/react-spectrum/pull/2993, mainly to avoid adding more complexity to useGridCell

This approach essentially makes the outer ListView div behave as if it is the gridcell so we can easily navigate to the focusable children but still applies gridrow specific labeling to it. The inner gridcell div becomes decorative mainly 

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

File: packages/@react-spectrum/list/src/ListView.tsx [hunk 1]
```diff
@@ -145,15 +146,13 @@ function ListView<T extends object>(props: ListViewProps<T>, ref: DOMRef<HTMLDiv
     selectionBehavior: props.selectionStyle === 'highlight' ? 'replace' : 'toggle'
   });
   let layout = useListLayout(state, props.density || 'regular');
-  let keyboardDelegate = useMemo(() => new GridKeyboardDelegate({
+  let keyboardDelegate = useMemo(() => new ListGridKeyboardDelegate({
```
Comment (LFDanLu) [inline packages/@react-spectrum/list/src/ListView.tsx:149]: Note that the `focusMode` of the gridState and the keyboard delegate is still `cell`, I think I could change it to `row` but it was just easier to go ahead and delete `onFocus` from `useGridRow` rather than from `useGridCell`. The `onFocus` from `useGridCell` has some additional logic merged in from `useSelectableItem `

File: packages/@react-spectrum/list/src/ListViewItem.tsx [hunk 2]
```diff
@@ -55,17 +55,28 @@ export function ListViewItem(props) {
     onAction: onAction ? () => onAction(item.key) : undefined,
     shouldSelectOnPressUp: isListDraggable
   }, state, rowRef);
```
Comment (LFDanLu) [inline packages/@react-spectrum/list/src/ListViewItem.tsx:57]: useGridRow retains the same `node: item` so selection and actions use the row key still

File: packages/@react-aria/list/package.json [hunk 3]
```diff
@@ -0,0 +1,32 @@
+{
+  "name": "@react-aria/list",
+  "version": "3.0.0-alpha.1",
+  "private": true,
```
Comment (LFDanLu) [inline packages/@react-aria/list/package.json:4]: I forget, should I directly modify this or will the version/private be updated upon publish?

File: packages/@react-aria/list/src/ListGridKeyboardDelegate.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {GridCollection} from '@react-types/grid';
+import {GridKeyboardDelegate, GridKeyboardDelegateOptions} from '@react-aria/grid';
+import {Key} from 'react';
+import {Rect} from '@react-stately/virtualizer';
+
+export class ListGridKeyboardDelegate<T> extends GridKeyboardDelegate<T, GridCollection<T>> {
```
Comment (LFDanLu) [inline packages/@react-aria/list/src/ListGridKeyboardDelegate.ts:18]: Open to name suggestions

File: packages/@react-aria/list/package.json [hunk 5]
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "@react-aria/list",
+  "version": "3.0.0-alpha.1",
```
Comment (LFDanLu) [inline packages/@react-aria/list/package.json:3]: Removed the private since it was failing lint (published package uses this). Lemme know if I should bump this version to 3.0.0

File: scripts/lint-packages.js [hunk 6]
```diff
@@ -106,29 +122,27 @@ for (let pkg of packages) {
...
-  if (scope === '@react-spectrum') {
-    let aria = `@react-aria/${basename}`;
-    let stately = `@react-stately/${basename}`;
-    let types = `@react-types/${basename}`;
 
+  let aria = `@react-aria/${basename}`;
+  let stately = `@react-stately/${basename}`;
+  let types = `@react-types/${basename}`;
+
+  if (scope === '@react-spectrum' && isDepUsed(aria, globSrc)) {
```
Comment (LFDanLu) [inline scripts/lint-packages.js:137]: Updated this lint step since it was complaining that I wasn't importing `react-stately/list` in `react-aria/list` even though `react-stately/list` wasn't used in `react-aria/list` at all. Now it checks that `@react-aria/${basename}`/`@react-stately/${basename}`/`@react-types/${basename}` is actually being imported somewhere in the package's files before asserting that it exists in the package.json.

Open to suggestions on the changes here and/or I've misunderstood the purpose of these checks

Comment (LFDanLu) [reply scripts/lint-packages.js:137]: dunno, I'd have to go through the aria packages we have to check if this is the first instance of this problem or not. If it isn't then I'm mystified why it was caught here specifically

File: packages/@react-spectrum/list/test/ListView.test.js [hunk 7]
```diff
@@ -166,40 +218,35 @@ describe('ListView', function () {
...
-        )}
-      </ListView>,
-      locale,
-      scale
-    );
+  it('should focus the row if the cell is pressed', function () {
+    let tree = renderList({selectionMode: 'single'});
+    let cell = within(getRow(tree, 'Bar')).getByRole('gridcell');
+    act(() => {
+      triggerPress(cell);
```
[the remark this answers is not in the document]
Comment (LFDanLu) [reply packages/@react-spectrum/list/test/ListView.test.js:232]: lemme try that

File: scripts/lint-packages.js (hunk 6 above)
Comment (LFDanLu) [reply scripts/lint-packages.js:137]: Ah right, so the reason why this wasn't a problem was because of this part of the check being passed to the `softAssert`: `!pkgNames[stately] || json.dependencies[stately]`. For a package like `@react-aria/accordion`, the first part will resolve to true since the `@react-stately/accordion` package doesn't exist (aka `pkgNames` doesn't contain `@react-stately/accordion`). However, for the newly introduced `@react-aria/list` package (ListView specific), `@react-stately/list` does exist so it expects it to be in the dependencies even though it isn't being used.

I could rename the listview package name to be `listview` instead of `list` which should side step this problem but it kinda breaks the previous naming conventions (e.g. TableView's package is `table`). Also I think that the update to this script still makes sense, doesn't feel like the previous logic was making a correct assumption.

Comment (snowystinger) [reply scripts/lint-packages.js:137]: yep, thanks for looking into that! i think when I wrote that I assumed naming parity
so maybe we should still reconsider the name of the package, however, the change you made also makes sense

**Comment by snowystinger** [comment]
Screen Shot 2022-04-13 at 2 31 10 PM
ok, this one is on my native monitor, top border seems to be too thick?

**Comment by LFDanLu** [comment]
@snowystinger yep, definitely too thick. Pre-existing issue though, should be cleared up by my [ListView UI update pr](https://github.com/adobe/react-spectrum/pull/2804)

**Review by snowystinger — APPROVED** [review]
In that case, tested on Chrome/Safari, looks good
I think there were two cases of naming needing to be resolved? Let's bring it up in a team meeting.

File: packages/@react-spectrum/list/src/ListViewItem.tsx [hunk 8]
```diff
@@ -118,8 +130,7 @@ export function ListViewItem(props) {
             }
           )
         }
-        ref={cellRef}
-        {...mergedProps}>
+        role="gridcell">
```
Comment (devongovett) [inline packages/@react-spectrum/list/src/ListViewItem.tsx:133]: We'll probably eventually want to move all this logic into a `useListItem` hook since it's no longer a pure grid. But doesn't have to be in this pr.

File: packages/@react-aria/list/src/ListGridKeyboardDelegate.ts (hunk 4 above)
Comment (devongovett) [reply packages/@react-aria/list/src/ListGridKeyboardDelegate.ts:18]: What are the differences between this and ListKeyboardDelegate now?

Comment (LFDanLu) [reply packages/@react-aria/list/src/ListGridKeyboardDelegate.ts:18]: ListKeyboardDelegate will skip over disabled items whereas this one doesn't. Also handles the `getKeyLeftOf`/`getKeyRightOf` that useGridCell calls which ListKeyboardDelegate doesn't

File: packages/@react-spectrum/list/src/ListViewItem.tsx (hunk 8 above)
Comment (LFDanLu) [reply packages/@react-spectrum/list/src/ListViewItem.tsx:133]: Ah, good idea, I'll handle in a followup PR once we settle on names and stuff

File: packages/@react-aria/list/src/ListGridKeyboardDelegate.ts (hunk 4 above)
Comment (LFDanLu) [reply packages/@react-aria/list/src/ListGridKeyboardDelegate.ts:18]: Sure I can go ahead and try modifying ListKeyboardDelegate instead

Comment (LFDanLu) [reply packages/@react-aria/list/src/ListGridKeyboardDelegate.ts]: I've done some poking around and have come up with https://github.com/adobe/react-spectrum/pull/3000/commits/f2ef4ef8e995f3e11f1520bfd35c9b450a74f86c, which leverages `ListLayout` as the keyboard delegate instead of the new `ListGridKeyboardDelegate`. I feel a bit iffy about it and it still has some finicky behavior that I can't resolve without bringing in some grid specific logic into ListLayout. 

Pain points:
- An effort has been made to make the tracked `focusedKey` always be a row key. However, due to the nature of `useGridCell`, it is possible for the `focusedKey` to be a cell key (e.g. if a user clicks on a row button), resulting in some broken behavior since the ListLayout wasn't designed to handle such a distinction. If I get rid of the `onFocus` from `useGridCell` that would fix the problem of `focusedKey` being set to a cell key but would then mean focus wouldn't update properly if the user clicks on said row buttons. Focus would also get highjacked from a menu button after closing the menu and returned onto the row since `useGridRow` doesn't have it's own dedicated `focus` func that handles these kinds of situations.  
- the list collection only has access to the `row` keys, not the `cell` keys. This makes using `ListKeyboardDelegate`/`ListLayout` tricky since they don't know what to do with the `cell` keys unless I modify those functions so they always try to grab the `row` key from the provided grid collection
- This mixture of using a grid collection with layouts/delegates that don't have a concept of `row` vs `cell` is pretty painful in general
- `ListKeyboarddDelegate` makes an assumption that the component isn't virtualized and uses query selectors to find row height information

I found the old approach w/ the new keyboard delegate to be a bit easier to navigate since it was extending off the GridKeyboardDelegate and I was able to ensure the `focusedKey` would always be the `cell` key meaning there were less pitfalls

File: packages/@react-spectrum/list/src/ListViewItem.tsx (hunk 2 above)
Comment (LFDanLu) [reply packages/@react-spectrum/list/src/ListViewItem.tsx:60]: outdated now

File: packages/@react-stately/layout/src/ListLayout.ts [hunk 9]
```diff
@@ -25,7 +25,8 @@ export type ListLayoutOptions<T> = {
   indentationForItem?: (collection: Collection<Node<T>>, key: Key) => number,
   collator?: Intl.Collator,
   loaderHeight?: number,
-  placeholderHeight?: number
+  placeholderHeight?: number,
+  allowDisabledKeyFocus?: boolean
```
Comment (LFDanLu) [inline packages/@react-stately/layout/src/ListLayout.ts:29]: New option for ListLayout, makes disabled rows focusable via keyboard navigation, defaults to false.

**Comment by devongovett** [comment]
One issue I see: In the dynamic items story, if you focus one of the buttons in the first row, tab out of the list view and shift tab back in, it focuses a different row rather than the first one. Didn't seem to happen previously. Seems like it's going to the last one that is visible.

**Comment by LFDanLu** [comment]
Strange, doesn't seem to be consistent either. Maybe something with useSelectableCollection and how focuses the last focusable element + GridCell behavior? I'll dig

**Comment by LFDanLu** [comment]
@devongovett think I've narrowed it down to it being specifically a story issue since tabbing out of the ListView moves you out of the iframe but I'll go into the details.

So the code that is causing the last visible row to be focused is https://github.com/adobe/react-spectrum/blob/a7642bfef6bf4de8af8a6ee9c098592c67ce2e04/packages/%40react-aria/grid/src/useGridCell.ts#L206-L208. Due to standard browser behavior, shift tabbing back into the ListView will always focus the last focusable element (aka the last button in the last row) before attempting to move focus to the tracked `focusedKey`. Typically `useGridCell`'s `onFocus` will ignore the focus event that happens if you are in keyboard modality, which for the most part works out for us when the user tabs/shift-tabs into the ListView since we have a document level keyboard listener that shift modality to "keyboard".
 
For this story however, tabbing out of the ListView will bring you out of the iframe. If you then move the pointer within the story iframe, modality will shift to "pointer". The subsequent Shift + Tab operation to bring focus back into the ListView happens outside of the iframe and thus the event listener responsible for shifting back to keyboard modality doesn't fire, resulting in `useGridCell` setting the `focusedKey` to the last rendered row since `isFocusVisible` resolves to false.

Here are some 100% reproduction steps (TableView story since it isn't just for ListView):
1. Navigate to https://reactspectrum.blob.core.windows.net/reactspectrum/2e88e5c9f2feb6aea265db1cfbdb3125cbff1699/storybook/index.html?path=/story/tableview--dynamic-with-selection
2. Tab into the table. Note what row is focused
3. Tab out of the table.
4. Move you mouse around in the story to shift modality to 'pointer'
5. Shift tab back into the table. Note that the row 7 checkbox is focused.

TLDR: This is pre-existing behavior and only happens if your focus leaves the document entirely. It won't typically happen if you have focusable elements preceding and following the ListView/TableView.
I can add input fields before and after each ListView in the stories if you'd like

File: packages/@react-spectrum/list/test/ListView.test.js [hunk 10]
```diff
@@ -236,12 +283,14 @@ describe('ListView', function () {
...
         it('should move focus to previous cell in RTL', function () {
-          let tree = renderListWithFocusables('ar-AE');
+          let tree = renderListWithFocusables({locale: 'ar-AE'});
           // Should move from button two to button one
-          let start = within(tree.getAllByRole('row')[0]).getAllByRole('button')[1];
-          let end = within(tree.getAllByRole('row')[0]).getAllByRole('button')[0];
-          act(() => start.focus());
+          let start = within(getRow(tree, 'Foo')).getAllByRole('button')[1];
+          let end = within(getRow(tree, 'Foo')).getAllByRole('button')[0];
+          // Need to press to set a modality, otherwise useSelectableCollection will think this is a tab operation
```
Comment (snowystinger) [inline packages/@react-spectrum/list/test/ListView.test.js:290]: have to wonder if `.focus` is the best way to do it anyways, I think it might make more sense to call `userEvent.tab()`, that'd be more representative of the real world

Comment (snowystinger) [reply packages/@react-spectrum/list/test/ListView.test.js:290]: for the other instances, not for this one, where clicking is the real life thing happening

File: packages/@react-spectrum/list/test/ListView.test.js [hunk 11]
```diff
@@ -288,42 +338,144 @@ describe('ListView', function () {
...
         expect(document.activeElement).toBe(start);
       });
 
       it('should move focus to above row', function () {
         let tree = renderListWithFocusables();
-        let start = getCell(tree, 'Bar');
-        let end = getCell(tree, 'Foo');
+        let start = getRow(tree, 'Bar');
+        let end = getRow(tree, 'Foo');
+        act(() => start.focus());
```
Comment (snowystinger) [inline packages/@react-spectrum/list/test/ListView.test.js:353]: like, this one should be `userEvent.tab()` i think

File: packages/@react-spectrum/list/src/ListViewItem.tsx [hunk 12]
```diff
@@ -32,47 +32,42 @@ export function ListViewItem(props) {
...
-  let allowsInteraction = state.selectionManager.selectionMode !== 'none' || onAction;
+  let allowsInteraction = state.selectionManager.selectionMode !== 'none' || hasActions;
   let isDisabled = !allowsInteraction || state.disabledKeys.has(item.key);
   let isDraggable = dragState?.isDraggable(item.key) && !isDisabled;
   let {hoverProps, isHovered} = useHover({isDisabled});
   let {pressProps, isPressed} = usePress({isDisabled});
-  let {rowProps} = useGridRow({
+
+  // We only make use of useGridCell here to allow for keyboard navigation to the focusable children of the row.
+  // The actual grid cell of the ListView is intert since we don't want to ever focus it to decrease screenreader
```
Comment (snowystinger) [inline packages/@react-spectrum/list/src/ListViewItem.tsx:53]: intert?

**Review by devongovett — DISMISSED** [review]

File: packages/@react-spectrum/list/test/ListView.test.js (hunk 11 above)
Comment (LFDanLu) [reply packages/@react-spectrum/list/test/ListView.test.js:353]: The start.focus was a easy shortcut to get focus onto a row that isn't the starting row. I can see about using a press instead

**Review by devongovett — APPROVED** [review]

**Review by snowystinger — APPROVED** [review]
